src/frontend/src/scheduler/hummock_snapshot_manager.rs

```diff
@@ -44,26 +46,43 @@ impl HummockSnapshotManager {
                 .await?;
             core_guard.is_outdated = false;
             core_guard.last_pinned = epoch;
-            core_guard.reference_number.insert(epoch, 1);
-        } else {
-            let last_pinned = core_guard.last_pinned;
-            *core_guard.reference_number.get_mut(&last_pinned).unwrap() += 1;
+            core_guard.epoch_to_query_ids.insert(epoch, HashSet::new());
         }
+        let last_pinned = core_guard.last_pinned;
+        core_guard
+            .epoch_to_query_ids
+            .get_mut(&last_pinned)
+            .unwrap()
+            .insert(query_id);
         Ok(core_guard.last_pinned)
     }
 
-    pub async fn unpin_snapshot(&self, epoch: u64) -> Result<()> {
-        let mut core_guard = self.core.lock().await;
-        let reference_number = core_guard.reference_number.get_mut(&epoch);
-        if let Some(reference_number) = reference_number {
-            *reference_number -= 1;
-            if *reference_number == 0 {
-                self.meta_client.unpin_snapshot(epoch).await?;
-                core_guard.reference_number.remove(&epoch);
-                if epoch == core_guard.last_pinned {
-                    core_guard.is_outdated = true;
+    pub async fn unpin_snapshot(&self, epoch: u64, query_id: &QueryId) -> Result<()> {
+        let local_count = async {
+            let mut core_guard = self.core.lock().await;
+            let query_ids = core_guard.epoch_to_query_ids.get_mut(&epoch);
+            if let Some(query_ids) = query_ids {
+                query_ids.remove(query_id);
+                if query_ids.is_empty() {
+                    core_guard.epoch_to_query_ids.remove(&epoch);
+                    if epoch == core_guard.last_pinned {
+                        core_guard.is_outdated = true;
+                    }
+                    return true;
                 }
             }
+            false
+        };
+        let need_to_request_meta = local_count.await;
+        if need_to_request_meta {
+            let meta_client = self.meta_client.clone();
+            tokio::spawn(async move {
+                let handle = tokio::spawn(async move { meta_client.unpin_snapshot(epoch).await });
+
+                if let Err(join_error) = handle.await && join_error.is_panic() {
+                    error!("Request meta to unpin snapshot panic {:?}!", join_error);
+                }
```

Comment:
The error handle here looks ok. In the future, I want to trigger unpin_snapshot every time fe receives a update. So if the unpin fails, it will be tried again next time received meta push.